pymc3/distributions/continuous.py

```diff
@@ -632,99 +673,63 @@ class TruncatedNormal(BoundedContinuous):
 
     """
 
-    def __init__(
-        self,
-        mu=0,
-        sigma=None,
-        tau=None,
-        lower=None,
-        upper=None,
-        transform="auto",
-        sd=None,
+    rv_op = truncated_normal
+    bound_args_indices = (2, 3)  # indexes for lower and upper args
+
+    @classmethod
+    def dist(
+        cls,
+        mu: Optional[Union[float, np.ndarray]] = None,
+        sigma: Optional[Union[float, np.ndarray]] = None,
+        tau: Optional[Union[float, np.ndarray]] = None,
+        sd: Optional[Union[float, np.ndarray]] = None,
+        lower: Optional[Union[float, np.ndarray]] = None,
+        upper: Optional[Union[float, np.ndarray]] = None,
+        transform: str = "auto",
         *args,
         **kwargs,
-    ):
-        if sd is not None:
-            sigma = sd
+    ) -> RandomVariable:
+        sigma = sd if sd is not None else sigma
         tau, sigma = get_tau_sigma(tau=tau, sigma=sigma)
-        self.sigma = self.sd = at.as_tensor_variable(sigma)
-        self.tau = at.as_tensor_variable(tau)
-        self.lower_check = at.as_tensor_variable(floatX(lower)) if lower is not None else lower
-        self.upper_check = at.as_tensor_variable(floatX(upper)) if upper is not None else upper
-        self.lower = (
-            at.as_tensor_variable(floatX(lower))
-            if lower is not None
-            else at.as_tensor_variable(-np.inf)
-        )
-        self.upper = (
-            at.as_tensor_variable(floatX(upper))
-            if upper is not None
-            else at.as_tensor_variable(np.inf)
-        )
-        self.mu = at.as_tensor_variable(floatX(mu))
-
-        if self.lower_check is None and self.upper_check is None:
-            self._defaultval = mu
-        elif self.lower_check is None and self.upper_check is not None:
-            self._defaultval = self.upper - 1.0
-        elif self.lower_check is not None and self.upper_check is None:
-            self._defaultval = self.lower + 1.0
-        else:
-            self._defaultval = (self.lower + self.upper) / 2
-
+        single_param_length = 0 if not sigma.shape else sigma.shape[0]
```

Comment:
I played a bit with different options and I cannot find a decent one supporting symbolic shapes.

If use a python if else I will be able to return a scalar when scalars as passed as inputs and an array otherwise. However I won't be able to use the symbolic shape (unless I eval it). eg:
```python 
def _handle_missing_bound(bound_value, expected_input_length, replacement_value): 
	if bound_value is not None: 
    	    return at.as_tensor_variable(floatX(bound_value)) 
	elif expected_input_length == 0: 
    	    return at.as_tensor_variable(floatX(replacement_value)) 
	else: 
    	    return at.repeat(at.as_tensor_variable(replacement_value), at.as_tensor_variable(expected_input_length))

```

If on the other hand I use at,switch I end up always having an array as output (an empy one when mu_shape is empty)
```python
switch(at.gt(mu_shape.size, at.as_tensor_variable(0)), at.repeat(at.as_tensor_variable(floatX(replacement_value)), mu_shape.size), replacement_value)
```

Alternatively I can use at.repeat directly, but again I end up always having an array as output
```python
repeat = at.switch(at.gt(mu_shape.size, at.as_tensor_variable(0)), mu_shape.shape[0], 1)
bound_value = at.repeat(at.as_tensor_variable(floatX(replacement_value)), repeat).eval()
```
None of the cases above seems a serious solution to me, at the same time I cannot think a decent symbolic approach supporting both scalars and arrays. I have fairly little Theano knowledge, so there may be an obvious solution to this

Reply:
I might have missed / forgot it, but why can't we pass a simple constant `-np.inf`. The RandomVariable should be able to broadcast it to a vector when needed.

Reply:
Thanks for stepping in and improving it.

> The RandomVariable should be able to broadcast it to a vector when needed.
Oh! I didn't know about it and I thought it was necessary to handle it in`.dist`. That's the reason why I wanted to use `np.repeat` /`at.repeat`